### ISSUE TITLE
train.py

### ISSUE THREAD

**Issue by irfanICMLL** [irfanICMLL/dvn#6]
# _*_ coding: utf-8
import os

import numpy as np
import tensorflow as tf

from utils.data_handle import save_weight, load_weight
from utils.image_process import prepare_label, inv_preprocess, decode_labels
from utils.image_reader import ImageReader

def convert_to_calculateloss(raw_output, num_classes, label_batch, isFull = False):
    if isFull:
        raw_groundtruth = tf.reshape(tf.squeeze(label_batch, squeeze_dims=[3]), [-1, ])
    else:
        label_proc = prepare_label(label_batch, raw_output.get_shape()[1:3],
                                num_classes=num_classes, one_hot=False)  # [batch_size, h, w]
        raw_groundtruth = tf.reshape(label_proc, [-1, ])

    raw_prediction = tf.reshape(raw_output, [-1, num_classes])

    indices = tf.squeeze(tf.where(tf.less_equal(raw_groundtruth, num_classes - 1)), 1)
    label = tf.cast(tf.gather(raw_groundtruth, indices), tf.int32)  # [?, ]
    logits = tf.gather(raw_prediction, indices)  # [?, num_classes]

    return label, logits

def train(args, dbargs):

    if args.train_on_16:
        from models.deeplabnet_s16 import DeepLabV2, DeepLabV3
        from models.resnet_s16 import ResNet38, ResNet101
        args.batch_size = 10
    elif args.train_on_4:
        from models.deeplabnet_s4 import DeepLabV2, DeepLabV3
        from models.resnet_s4 import ResNet38, ResNet101
        args.batch_size = 1
    else:
        from models.deeplabnet_s8 import DeepLabV2, DeepLabV3
        from models.resnet_s8 import ResNet38, ResNet101
        args.batch_size = 3

    def choose_model(model_name, base_model, image_batch):
        def choose_base_model(base_model):
            if base_model == 'resnet38':
                net = ResNet38(inputs={'data': image_batch}).terminals[-1]
            elif base_model == 'resnet101':
                net = ResNet101(inputs={'data': image_batch}).terminals[-1]
            return net

        net = choose_base_model(base_model)
        if model_name == 'deeplabv2':
            net = DeepLabV2(inputs={net.op.name: net})
        elif model_name == 'deeplabv3':
            net = DeepLabV3(inputs={net.op.name: net})

        return net

    ## set hyparameter
    img_mean = np.array((104.00698793, 116.66876762, 122.67891434), dtype=np.float32)
    tf.set_random_seed(args.random_seed)
    coord = tf.train.Coordinator()

    ## load data
    with tf.name_scope("create_inputs"):
        reader = ImageReader(
            args.data_dir,
            args.img_size,
            args.random_scale,
            args.random_mirror,
            args.random_crop,
            args.ignore_label,
            args.is_val,
            img_mean,
            coord)
        image_batch, label_batch = reader.dequeue(args.batch_size)

    ## load model
    net = choose_model(args.model_name, args.base_model, image_batch)
    raw_output = net.get_output()

    predict_batch = net.topredict(raw_output, tf.shape(image_batch)[1:3])

    # 1/16 * 513
    label, logits = convert_to_calculateloss(raw_output, args.num_classes, label_batch)

    # 1 * 513
    labels_full, logits_full = convert_to_calculateloss(tf.image.resize_bilinear(raw_output,tf.shape(label_batch)[1:3]), args.num_classes, label_batch, isFull = True)
    predict_label = tf.argmax(logits, axis=1)
    pridict_full_label = tf.argmax(logits_full, axis = 1)
    print("Model load completed!")

    ## get all kinds of variables list
    def printV(var_list):
        for var in var_list:
            print(var)
        print("--------------------------------")

    basemodel_var = [v for v in tf.global_variables() if 'fc' not in v.name]  # restore pretrained model
    if args.is_training:
        all_trainable_var = [v for v in tf.trainable_variables()]
        print('batch normalization parameters are trained with decay = 0.9997')
    else:
        all_trainable_var = [v for v in tf.trainable_variables() if 'beta' not in v.name or 'gamma' not in v.name]
        print('batch normalization parameters are freezed')
    conv_trainable_var = [v for v in all_trainable_var if 'fc' not in v.name]
    fc_trainable_var = [v for v in all_trainable_var if 'fc' in v.name]
    fc_trainable_w_var = [v for v in fc_trainable_var if 'weight' in v.name]
    fc_trainable_b_var = [v for v in fc_trainable_var if 'biases' in v.name]

    ## set loss
    loss = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(labels=label, logits=logits))
    #l2_loss = [args.weight_decay * tf.nn.l2_loss(w) for w in fc_trainable_w_var]
    #loss = tf.add(loss, tf.add_n(l2_loss))
    loss_var, loss_op = tf.metrics.mean(loss)

    # output_stride: 16 miou
    iou_var, iou_op = tf.metrics.mean_iou(label, predict_label, args.num_classes)

    # output_stride: 1 miou
    iou_full_var, iou_full_op = tf.metrics.mean_iou(labels_full, pridict_full_label, args.num_classes)

    accuracy_var, acc_op = tf.metrics.accuracy(label, predict_label)
    metrics_op = tf.group(loss_op, iou_op, iou_full_op, acc_op)

    ## set optimizer
    iterstep = tf.placeholder(dtype=tf.float32, shape=[], name='iteration_step')

    base_lr = tf.constant(args.learning_rate, dtype=tf.float32, shape=[])
    lr = tf.scalar_mul(base_lr,
                       tf.pow((1 - iterstep / args.num_steps), args.power))  # learning rate reduce with the time

    train_basemodel_op = tf.train.MomentumOptimizer(learning_rate=lr,
                                                    momentum=args.momentum).minimize(loss,
                                                                                     var_list=conv_trainable_var)
    train_fc_op = tf.train.MomentumOptimizer(learning_rate=lr,
                                             momentum=args.momentum).minimize(loss,
                                                                              var_list=fc_trainable_var)
    train_fc_w_op = tf.train.MomentumOptimizer(learning_rate=lr * 10,
                                               momentum=args.momentum).minimize(loss,
                                                                                var_list=fc_trainable_w_var)
    train_fc_b_op = tf.train.MomentumOptimizer(learning_rate=lr * 20,
                                               momentum=args.momentum).minimize(loss,
                                                                                var_list=fc_trainable_b_var)
    train_all_op = tf.group(train_basemodel_op, train_fc_op)

    # train_all_op = tf.group(train_basemodel_op, train_fc_w_op, train_fc_b_op)
    # train_fc_op = tf.group(train_fc_w_op, train_fc_b_op)

    ## set summary
    vs_image = tf.py_func(inv_preprocess, [image_batch, args.save_num_images, img_mean], tf.uint8)
    vs_label = tf.py_func(decode_labels, [label_batch, args.save_num_images, args.num_classes], tf.uint8)
    vs_predict = tf.py_func(decode_labels, [predict_batch, args.save_num_images, args.num_classes], tf.uint8)
    tf.summary.image(name='image collection_train', tensor=tf.concat(axis=2, values=[vs_image, vs_label, vs_predict]),
                     max_outputs=args.save_num_images)
    tf.summary.scalar('loss_train', loss_var)
    tf.summary.scalar('iou_belif_train', iou_var)
    tf.summary.scalar('iou_full_train', iou_full_var)
    tf.summary.scalar('accuracy_train', accuracy_var)

    summary_op = tf.summary.merge_all()
    summary_writer = tf.summary.FileWriter(dbargs['log_dir'], graph=tf.get_default_graph(), max_queue=5)

    ## set session
    print("GPU")
    os.system("echo $CUDA_VISIBLE_DEVICES")

    sess = tf.Session()
    global_init = tf.global_variables_initializer()
    local_init = tf.local_variables_initializer()
    sess.run(global_init)
    sess.run(local_init)

    ## set saver
    saver_g = tf.train.Saver(var_list=tf.global_variables(), max_to_keep=100)
    trained_step = 0
    if os.path.exists(dbargs['restore_from'] + 'checkpoint'):
        trained_step = load_weight(dbargs['restore_from'], saver_g, sess)
    else:
        saver_b = tf.train.Saver(var_list=basemodel_var)
        load_weight(dbargs['baseweight_from'], saver_b, sess)

    with open(dbargs['restore_from']+'info.me', 'w') as info:
        for k, v in vars(args).items():
            info.write('%s: %s\n' % (str(k), str(v)))

    threads = tf.train.start_queue_runners(sess, coord)
    print("start training")

    ## start training
    for step in range(args.num_steps):
        if args.from_epoch_0:
            now_step = step
        else:
            now_step = int(trained_step) + step if trained_step is not None else step
        if now_step >= args.num_steps:
            break
        feed_dict = {iterstep: now_step}
        label_batch_, losses, lrate, iou, _, _ = sess.run([label_batch, loss_var, lr, iou_var, train_all_op, metrics_op], feed_dict)
        if step % args.save_pred_every == 0:
            save_weight(dbargs['restore_from'], saver_g, sess, now_step)

        if step % 50 == 0:
            print('step:{}\tlr {}\tloss = {}\tious:{}'.format(now_step, lrate, losses, iou))
            summary_str = sess.run(summary_op, feed_dict)
            summary_writer.add_summary(summary_str, now_step)
            sess.run(local_init)

    ## end training
    coord.request_stop()
    coord.join(threads)
